src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+            tail.next = node;
+            node.prev = tail;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        throwIndexOutOfBoundsAdd(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> current = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        current.prev = (current.prev != null) ? current.prev.next = newNode : (head = newNode);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> current = findNodeByIndex(index);
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        throwIndexOutOfBounds(index);
+        Node<T> current = findNodeByIndex(index);
+        T oldValue = current.item;
+        if (index == 0) {
+            head.item = value;
+        } else {
+            current.item = value;
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        throwIndexOutOfBounds(index);
+        Node<T> current = findNodeByIndex(index);
+        return unlike(current);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (object == null) {
+                if (current.item == null) {
+                    unlike(current);
+                    return true;
+                }
+            } else {
+                if (object.equals(current.item)) {
+                    unlike(current);
+                    return true;
+                }
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+
+        public Node(T item) {
+            this.item = item;
+        }
+    }
+
+    private void throwIndexOutOfBoundsAdd(int index) throws IndexOutOfBoundsException {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds.");
+        }
+    }
+
+    private void throwIndexOutOfBounds(int index) throws IndexOutOfBoundsException {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds.");
+        }
+    }
```

Comment:
The method name 'throwIndexOutOfBounds' is not very informative. It would be better to rename it to something like 'checkIndex' to make its purpose clearer.